packages/react-events/src/dom/Tap.js

```diff
@@ -0,0 +1,664 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+
+import type {
+  ReactDOMResponderContext,
+  ReactDOMResponderEvent,
+  PointerType,
+} from 'shared/ReactDOMTypes';
+import type {ReactEventResponderListener} from 'shared/ReactTypes';
+
+import React from 'react';
+import {
+  buttonsEnum,
+  hasPointerEvents,
+  isMac,
+  dispatchDiscreteEvent,
+  dispatchUserBlockingEvent,
+} from './shared';
+
+type TapProps = {|
+  disabled: boolean,
+  preventDefault: boolean,
+  onTapCancel: (e: TapEvent) => void,
+  onTapChange: boolean => void,
+  onTapEnd: (e: TapEvent) => void,
+  onTapStart: (e: TapEvent) => void,
+  onTapUpdate: (e: TapEvent) => void,
+|};
+
+type TapState = {
+  activePointerId: null | number,
+  buttons: 0 | 1 | 4,
+  gestureState: TapGestureState,
+  ignoreEmulatedEvents: boolean,
+  isActive: boolean,
+  pointerType: PointerType,
+  responderTarget: null | Element,
+  rootEvents: null | Array<string>,
+  shouldPreventClick: boolean,
+};
+
+type TapEventType =
+  | 'tap-cancel'
+  | 'tap-change'
+  | 'tap-end'
+  | 'tap-start'
+  | 'tap-update';
+
+type TapGestureState = {|
+  altKey: boolean,
+  buttons: 0 | 1 | 4,
+  ctrlKey: boolean,
+  height: number,
+  metaKey: boolean,
+  pageX: number,
+  pageY: number,
+  pointerType: PointerType,
+  pressure: number,
+  screenX: number,
+  screenY: number,
+  shiftKey: boolean,
+  tangentialPressure: number,
+  target: null | Element,
+  tiltX: number,
+  tiltY: number,
+  timeStamp: number,
+  twist: number,
+  width: number,
+  x: number,
+  y: number,
+|};
+
+type TapEvent = {|
+  ...TapGestureState,
+  type: TapEventType,
+|};
+
+/**
+ * Native event dependencies
+ */
+
+const targetEventTypes = hasPointerEvents
+  ? ['pointerdown']
+  : ['mousedown', 'touchstart'];
+
+const rootEventTypes = hasPointerEvents
+  ? [
+      'click_active',
+      'contextmenu',
+      'pointerup',
+      'pointermove',
+      'pointercancel',
+      'scroll',
+    ]
+  : [
+      'click_active',
+      'contextmenu',
+      'mouseup',
+      'mousemove',
+      'dragstart',
+      'touchend',
+      'touchmove',
+      'touchcancel',
+      'scroll',
+    ];
+
+/**
+ * Responder and gesture state
+ */
+
+function createInitialState(): TapState {
+  return {
+    activePointerId: null,
+    buttons: 0,
+    ignoreEmulatedEvents: false,
+    isActive: false,
+    pointerType: '',
+    responderTarget: null,
+    rootEvents: null,
+    shouldPreventClick: true,
+    gestureState: {
+      altKey: false,
+      buttons: 0,
+      ctrlKey: false,
+      height: 1,
+      metaKey: false,
+      pageX: 0,
+      pageY: 0,
+      pointerType: '',
+      pressure: 0,
+      screenX: 0,
+      screenY: 0,
+      shiftKey: false,
+      tangentialPressure: 0,
+      target: document.createElement('div'),
```

Comment:
Why are we creating an element here? Also the `document` might not be the right one, as React works across multiple documents.

Reply:
This can just be `null` now that I've given up on not having all these ghost `null` types floating around